src/proc/merge-filter.cpp

```diff
@@ -0,0 +1,238 @@
+        if (depth_sequ_size == 0)
+            return false;
+
+        return true;
+    }
+
+
+    rs2::frame merge_filter::process_frame(const rs2::frame_source& source, const rs2::frame& f)
+    {
+        // steps:
+        // 1. get depth frame from incoming frameset
+        // 2. add the frameset to vector of framesets
+        // 3. check if size of this vector is at least 2 (if not - return latest merge frame)
+        // 4. pop out both framesets from the vector
+        // 5. apply merge algo
+        // 6. save merge frame as latest merge frame
+        // 7. return the merge frame
+
+        // 1. get depth frame from incoming frameset
+        auto fs = f.as<rs2::frameset>();
+        auto depth_frame = fs.get_depth_frame();        
+
+        // 2. add the frameset to vector of framesets
+        int depth_sequ_size = depth_frame.get_frame_metadata(RS2_FRAME_METADATA_SUBPRESET_SEQUENCE_SIZE);
+        if (depth_sequ_size > 2)
+            LOG_WARNING("merge_filter::process_frame(...) failed! sequence size must be 2 for merging pb.");
+
+        int depth_sequ_id = depth_frame.get_frame_metadata(RS2_FRAME_METADATA_SUBPRESET_SEQUENCE_ID);
+        
+        // condition added to ensure that frames are saved in the right order
+        // to prevent for example the saving of frame with sequence id 1 before
+        // saving frame of sequence id 0
+        // so that the merging with be deterministic - always done with frame n and n+1
+        // with frame n as basis
+        if (_framesets.size() == depth_sequ_id)
+        {
+            _framesets[depth_sequ_id] = fs;
+        }
+
+        // 3. check if size of this vector is at least 2 (if not - return latest merge frame)
+        if (_framesets.size() >= 2)
+        {
+            // 4. pop out both framesets from the vector
+            rs2::frameset fs_0 = _framesets[0];
+            rs2::frameset fs_1 = _framesets[1];
+            _framesets.clear();
+
+            bool use_ir = false;
+            if (check_frames_mergeability(fs_0, fs_1, use_ir))
+            {
+                // 5. apply merge algo
+                rs2::frame new_frame = merging_algorithm(source, fs_0, fs_1, use_ir);
+                if (new_frame)
+                {
+                    // 6. save merge frame as latest merge frame
+                    _depth_merged_frame = new_frame;
+                }
+            }
+        }
+
+        // 7. return the merge frame
+        if (_depth_merged_frame)
+        {
+            // discard saved merged frame if it is too old
+            auto fps = _depth_merged_frame.get_profile().fps();
+
+            auto input_frame_ts = f.get_frame_metadata(RS2_FRAME_METADATA_FRAME_TIMESTAMP);
+            auto depth_merged_frame_ts = _depth_merged_frame.get_frame_metadata(RS2_FRAME_METADATA_FRAME_TIMESTAMP);
+            auto delta_ts = input_frame_ts - depth_merged_frame_ts; // equal to 0 if frame created at this iteration, else it is positive
```

Comment:
What happens if the diff is negative? The new frame arriving can have lower TS as well.
Also  - is there an advantage to compare by TS rather than frame number?

Reply:
How can it be negative - how can the arriving frame have a ts lower than the previously saved merged frame?
If I compare only frame counter, I may fail in case of a test which starts pipe, then stops it and restarts it. The frame counter will be reset to zero and it might be used as depth merged frame at the beginning of the pipe restart.
This is actually the case of bug DSO-15644.

Reply:
TBD - Let's talk about this issue

Reply:
Done